File: backend/document-parser.js
import fs from 'fs';
import path from 'path';
import pdfParse from 'pdf-parse';
import mammoth from 'mammoth';

const AREA_KEYWORDS = {
  DWH: ['data warehouse', 'dwh', 'etl', 'data mart', 'staging', 'dimension', 'fact table'],
  MTII: ['mtii', 'market data', 'trading', 'instrument', 'pricing', 'valuation'],
  "Moody's": ['moody', 'rating', 'credit', 'analytics', 'risk assessment']
};

const COMPONENT_PATTERNS = {
  file_extracts: [/file extract/gi, /data extract/gi, /export file/gi, /feed/gi],
  data_models: [/data model/gi, /logical model/gi, /physical model/gi, /entity model/gi],
  tables: [/(\d+)\s*table/gi, /table[s]?\s*[:=]?\s*(\d+)/gi, /create table/gi],
  fields: [/(\d+)\s*field/gi, /field[s]?\s*[:=]?\s*(\d+)/gi, /(\d+)\s*column/gi],
  packages: [/package/gi, /stored procedure/gi, /pl\/sql/gi, /function/gi]
};

const COMPLEXITY_INDICATORS = {
  high: ['complex', 'critical', 'multiple systems', 'cross-functional', 'regulatory', 'compliance'],
  low: ['simple', 'straightforward', 'minor', 'small change', 'single']
};

async function extractTextFromPdf(filePath) {
  const dataBuffer = fs.readFileSync(filePath);
  const data = await pdfParse(dataBuffer);
  return data.text;
}

async function extractTextFromDocx(filePath) {
  const result = await mammoth.extractRawText({ path: filePath });
  return result.value;
}

export async function parseDocument(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  let text = '';

  try {
    if (ext === '.pdf') {
      text = await extractTextFromPdf(filePath);
    } else if (ext === '.docx') {
      text = await extractTextFromDocx(filePath);
    } else {
      throw new Error(`Unsupported file format: ${ext}`);
    }
  } catch (error) {
    console.error('Error parsing document:', error);
    return {
      filename: path.basename(filePath),
      extracted_text: '',
      identified_areas: [],
      identified_components: {},
      complexity: 'normal',
      complexity_indicators: [],
      suggested_impacts: { dwh: false, mtii: false, moodys: false }
    };
  }

  const textLower = text.toLowerCase();

  // Identify areas
  const identified_areas = [];
  for (const [area, keywords] of Object.entries(AREA_KEYWORDS)) {
    for (const keyword of keywords) {
      if (textLower.includes(keyword)) {
        if (!identified_areas.includes(area)) {
          identified_areas.push(area);
        }
        break;
      }
    }
  }

  // Count components
  const identified_components = {};
  for (const [component, patterns] of Object.entries(COMPONENT_PATTERNS)) {
    let count = 0;
    for (const pattern of patterns) {
      const matches = text.match(pattern) || [];
      count += matches.length;
      
      // Try to extract numbers
      for (const match of matches) {
        const nums = match.match(/\d+/);
        if (nums) {
          count = Math.max(count, parseInt(nums[0]));
        }
      }
    }
    identified_components[component] = count > 0 ? Math.max(1, count) : 0;
  }

  // Detect complexity
  let highScore = 0;
  let lowScore = 0;
  const complexity_indicators = [];

  for (const indicator of COMPLEXITY_INDICATORS.high) {
    if (textLower.includes(indicator)) {
      highScore++;
      complexity_indicators.push(`high: ${indicator}`);
    }
  }
  for (const indicator of COMPLEXITY_INDICATORS.low) {
    if (textLower.includes(indicator)) {
      lowScore++;
      complexity_indicators.push(`low: ${indicator}`);
    }
  }

  let complexity = 'normal';
  if (highScore > lowScore) complexity = 'complex';
  else if (lowScore > highScore) complexity = 'low';

  return {
    filename: path.basename(filePath),
    extracted_text: text.substring(0, 2000),
    identified_areas,
    identified_components,
    complexity,
    complexity_indicators,
    suggested_impacts: {
      dwh: identified_areas.includes('DWH'),
      mtii: identified_areas.includes('MTII'),
      moodys: identified_areas.includes("Moody's")
    }
  };
}

File: backend-python/document_parser.py
import os
import re

def parse_document(filepath):
    """Parse PDF or DOCX document and extract relevant information."""
    ext = os.path.splitext(filepath)[1].lower()
    
    if ext == '.pdf':
        text = extract_pdf_text(filepath)
    elif ext in ['.docx', '.doc']:
        text = extract_docx_text(filepath)
    else:
        raise ValueError(f'Unsupported file type: {ext}')
    
    return analyze_text(text)

def extract_pdf_text(filepath):
    """Extract text from PDF file."""
    try:
        from PyPDF2 import PdfReader
        reader = PdfReader(filepath)
        text = ''
        for page in reader.pages:
            text += page.extract_text() or ''
        return text
    except Exception as e:
        raise Exception(f'Error reading PDF: {str(e)}')

def extract_docx_text(filepath):
    """Extract text from DOCX file."""
    try:
        from docx import Document
        doc = Document(filepath)
        text = '\n'.join([para.text for para in doc.paragraphs])
        return text
    except Exception as e:
        raise Exception(f'Error reading DOCX: {str(e)}')

def analyze_text(text):
    """Analyze document text to extract impacted areas and component counts."""
    text_lower = text.lower()
    
    # Detect impacted areas
    impacted_areas = []
    
    area_keywords = {
        'DWH': ['dwh', 'data warehouse', 'datawarehouse', 'warehouse'],
        'MTII': ['mtii', 'mt2', 'market risk', 'trading'],
        'Moodys': ['moody', 'moodys', "moody's", 'rating', 'credit risk']
    }
    
    for area, keywords in area_keywords.items():
        if any(kw in text_lower for kw in keywords):
            impacted_areas.append(area)
    
    # Extract component counts using patterns
    component_counts = {
        'tables': 0,
        'fields': 0,
        'packages': 0,
        'reports': 0,
        'interfaces': 0
    }
    
    # Pattern matching for counts
    patterns = {
        'tables': [
            r'(\d+)\s*(?:new\s+)?tables?',
            r'tables?\s*[:\-]?\s*(\d+)',
            r'(\d+)\s*database\s+tables?'
        ],
        'fields': [
            r'(\d+)\s*(?:new\s+)?fields?',
            r'fields?\s*[:\-]?\s*(\d+)',
            r'(\d+)\s*(?:data\s+)?columns?'
        ],
        'packages': [
            r'(\d+)\s*(?:new\s+)?packages?',
            r'packages?\s*[:\-]?\s*(\d+)',
            r'(\d+)\s*(?:pl/?sql\s+)?modules?'
        ],
        'reports': [
            r'(\d+)\s*(?:new\s+)?reports?',
            r'reports?\s*[:\-]?\s*(\d+)'
        ],
        'interfaces': [
            r'(\d+)\s*(?:new\s+)?interfaces?',
            r'interfaces?\s*[:\-]?\s*(\d+)',
            r'(\d+)\s*(?:api\s+)?endpoints?'
        ]
    }
    
    for component, pattern_list in patterns.items():
        for pattern in pattern_list:
            matches = re.findall(pattern, text_lower)
            if matches:
                component_counts[component] = max(component_counts[component], max(int(m) for m in matches))
    
    # Detect complexity
    complexity = 'medium'
    
    high_complexity_indicators = [
        'complex', 'complicated', 'challenging', 'critical',
        'high risk', 'significant changes', 'major', 'extensive'
    ]
    
    low_complexity_indicators = [
        'simple', 'straightforward', 'minor', 'small',
        'low risk', 'minimal changes', 'basic'
    ]
    
    high_count = sum(1 for indicator in high_complexity_indicators if indicator in text_lower)
    low_count = sum(1 for indicator in low_complexity_indicators if indicator in text_lower)
    
    if high_count > low_count + 2:
        complexity = 'high'
    elif low_count > high_count + 2:
        complexity = 'low'
    
    return {
        'impacted_areas': impacted_areas,
        'component_counts': component_counts,
        'complexity': complexity,
        'text_preview': text[:500] + '...' if len(text) > 500 else text
    }
